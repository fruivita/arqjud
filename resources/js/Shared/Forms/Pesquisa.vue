<!--
    Componente de pesquisa.

    @see https://vuejs.org/guide/introduction.html
    @see https://tailwindcss.com/docs
    @see https://inertiajs.com/
 -->

<script>
export default { inheritAttrs: false };
</script>

<script setup>
import ContadorCaracteres from '@/Shared/Forms/ContadorCaracteres.vue';
import MensagemErro from '@/Shared/Forms/MensagemErro.vue';
import Icone from '@/Shared/Icones/Icone.vue';
import { usePage } from '@inertiajs/inertia-vue3';
import { computed, ref } from 'vue';

const props = defineProps({
    modelValue: { type: String },
    maxlength: { type: [Number, String], default: 50 },
});

const exibirContador = ref(false);

const erro = computed(() => usePage().props.value.errors.termo);

const emit = defineEmits(['update:modelValue']);

const onInput = (event) => {
    emit('update:modelValue', event.target.value);
};
</script>

<template>
    <div class="space-y-1 md:mx-auto md:w-2/4">
        <div
            :class="{
                'text-primaria-900 dark:text-secundaria-50': !erro,
                'text-red-900': erro,
            }"
            class="relative text-primaria-900"
        >
            <div class="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3">
                <Icone nome="search" />
            </div>

            <input
                :class="{
                    'border-primaria-500 bg-primaria-100 ring-primaria-500 dark:border-secundaria-600 dark:bg-secundaria-800 dark:ring-secundaria-500':
                        !erro,
                    'border-red-500 bg-red-100 ring-red-500': erro,
                }"
                :maxlength="maxlength"
                :value="modelValue"
                @blur="exibirContador = false"
                @focus="exibirContador = true"
                @input="onInput"
                v-bind="$attrs"
                class="w-full rounded-lg border py-2 pl-12 pr-20 outline-none focus:ring"
                autocomplete="off"
                type="text"
            />

            <ContadorCaracteres
                v-show="exibirContador"
                :maxlength="maxlength"
                :texto="modelValue"
            />
        </div>

        <MensagemErro v-if="erro" :erro="erro" />
    </div>
</template>
