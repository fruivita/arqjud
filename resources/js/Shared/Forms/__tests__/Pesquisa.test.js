/**
 * Testes para o componente Pesquisa.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import Pesquisa from '@/Shared/Forms/Pesquisa.vue';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test, vi } from 'vitest';

vi.stubGlobal('_dados', {});

vi.mock('@inertiajs/inertia-vue3', () => ({
    __esModule: true,
    ...vi.importActual('@inertiajs/inertia-vue3'),
    usePage: () => ({
        props: {
            value: {
                errors: {},
            },
        },
    }),
}));

let mountFunction;

beforeEach(() => {
    mountFunction = (options = {}) => {
        return mount(Pesquisa, { ...options });
    };
});

describe('Pesquisa', () => {
    test('propriedades estÃ£o definidas', () => {
        expect(Pesquisa.props).toMatchObject({
            modelValue: { type: String },
            maxlength: { type: [Number, String], default: 50 },
        });
    });

    test('renderiza o componente respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { modelValue: 'foo', maxlength: 30 },
            }).html()
        ).toMatchSnapshot();
    });

    test('atributos extras renderizados corretamente respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { modelValue: 'foo', maxlength: 30, extra: 'loren' },
            }).html()
        ).toMatchSnapshot();
    });

    test('exibe o componente filho ContadorCaracteres quando o input recebe foco', async () => {
        document.body.innerHTML = `
        <div>
            <div id="app"></div>
        </div>
        `;

        const wrapper = mountFunction({
            props: { modelValue: 'foo', maxlength: 30 },
            attachTo: document.getElementById('app'),
        });

        const contadorCaracteres = wrapper.findComponent({ name: 'ContadorCaracteres' });
        const input = wrapper.get('input');

        expect(contadorCaracteres.isVisible()).toBe(false);
        await input.trigger('focus');
        expect(contadorCaracteres.isVisible()).toBe(true);
        await input.trigger('blur');
        expect(contadorCaracteres.isVisible()).toBe(false);
    });

    test('atualiza o v-model ao alterar o valor do elemento', () => {
        const parent = mount({
            data: () => {
                return { foo: 'bar' };
            },
            template: '<div> <Pesquisa v-model="foo" /> </div>',
            components: { Pesquisa: Pesquisa },
        });

        const input = parent.find('input');

        input.element.value = 'bar baz';
        input.trigger('input');

        expect(parent.vm.foo).toBe('bar baz');
    });

    test('dispara o evento onInput', () => {
        const wrapper = mountFunction({
            props: { modelValue: '' },
        });

        const event = { target: { value: 'foo' } };

        wrapper.vm.onInput(event);

        expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['foo']);
    });
});
