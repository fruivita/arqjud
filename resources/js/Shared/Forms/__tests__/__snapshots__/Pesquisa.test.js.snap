// Vitest Snapshot v1

exports[`Pesquisa > atributos extras renderizados corretamente respeitando o snapshot 1`] = `
"<div class=\\"space-y-1 md:mx-auto md:w-2/4\\">
  <div class=\\"text-primaria-900 dark:text-secundaria-50 relative text-primaria-900\\">
    <div class=\\"pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path d=\\"M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z\\"></path>
      </svg></div><input class=\\"border-primaria-500 bg-primaria-100 ring-primaria-500 dark:border-secundaria-600 dark:bg-secundaria-800 dark:ring-secundaria-500 w-full rounded-lg border py-2 pl-12 pr-20 outline-none focus:ring\\" maxlength=\\"30\\" extra=\\"loren\\" autocomplete=\\"off\\" type=\\"text\\">
    <div class=\\"absolute inset-y-0 right-0 flex items-center pr-3 text-sm\\" style=\\"display: none;\\"><span dusk=\\"contador\\">3</span>/<span dusk=\\"maxlength\\">30</span></div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`Pesquisa > renderiza o componente respeitando o snapshot 1`] = `
"<div class=\\"space-y-1 md:mx-auto md:w-2/4\\">
  <div class=\\"text-primaria-900 dark:text-secundaria-50 relative text-primaria-900\\">
    <div class=\\"pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3\\"><svg class=\\"w-6 h-6\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
        <path d=\\"M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z\\"></path>
      </svg></div><input class=\\"border-primaria-500 bg-primaria-100 ring-primaria-500 dark:border-secundaria-600 dark:bg-secundaria-800 dark:ring-secundaria-500 w-full rounded-lg border py-2 pl-12 pr-20 outline-none focus:ring\\" maxlength=\\"30\\" autocomplete=\\"off\\" type=\\"text\\">
    <div class=\\"absolute inset-y-0 right-0 flex items-center pr-3 text-sm\\" style=\\"display: none;\\"><span dusk=\\"contador\\">3</span>/<span dusk=\\"maxlength\\">30</span></div>
  </div>
  <!--v-if-->
</div>"
`;
